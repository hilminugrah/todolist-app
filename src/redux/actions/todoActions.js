import axios from 'axios';

export const FETCH_TODOS_REQUEST = 'FETCH_TODOS_REQUEST';
export const FETCH_TODOS_SUCCESS = 'FETCH_TODOS_SUCCESS';
export const FETCH_TODOS_FAILURE = 'FETCH_TODOS_FAILURE';
export const TOGGLE_TODO = 'TOGGLE_TODO';
export const DELETE_TODO = 'DELETE_TODO';
export const ADD_TODO = 'ADD_TODO';


export const fetchTodos = () => async (dispatch) => {
  dispatch({ type: FETCH_TODOS_REQUEST });
  try {
    const res = await axios.get('https://jsonplaceholder.typicode.com/todos?_limit=10');
    dispatch({ type: FETCH_TODOS_SUCCESS, payload: res.data });
  } catch (error) {
    dispatch({ type: FETCH_TODOS_FAILURE, payload: error.message });
  }
};


export const toggleTodo = (id, completed) => async (dispatch) => {
  await axios.patch(`https://jsonplaceholder.typicode.com/todos/${id}`, { completed });
  dispatch({ type: TOGGLE_TODO, payload: { id, completed } });
};


export const deleteTodo = (id) => async (dispatch) => {
  await axios.delete(`https://jsonplaceholder.typicode.com/todos/${id}`);
  dispatch({ type: DELETE_TODO, payload: id });
};


export const addTodo = (title) => async (dispatch) => {
  const res = await axios.post('https://jsonplaceholder.typicode.com/todos', {
    title,
    completed: false,
  });
  dispatch({ type: ADD_TODO, payload: res.data });
};
