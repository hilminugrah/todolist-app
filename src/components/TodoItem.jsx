import React from 'react';
import { useDispatch } from 'react-redux';
import { toggleTodo, deleteTodo } from '../redux/actions/todoActions';
import { Trash2 } from 'lucide-react';

function TodoItem({ todo }) {
  const dispatch = useDispatch();

  const handleToggle = () => {
    dispatch(toggleTodo(todo.id, !todo.completed));
  };

  const handleDelete = () => {
    dispatch(deleteTodo(todo.id));
  };

  return (
    <li
      className={`flex justify-between items-center p-4 rounded-lg border border-gray-500 shadow-md ${
        todo.completed ? 'bg-gray-300' : 'bg-white'
      }`}
    >
      <div className="flex items-center space-x-3 relative">
        <input
          type="checkbox"
          checked={todo.completed}
          onChange={handleToggle}
          className={`
            w-5 h-5 appearance-none rounded-full border border-black
            checked:bg-purple-800 checked:border-purple-800
            relative transition duration-200
            after:content-[''] checked:after:content-['âœ”']
            after:absolute after:top-[1px] after:left-[5px]
            checked:after:text-white checked:after:font-bold checked:after:text-sm
          `}
        />
        <span className={todo.completed ? 'line-through text-gray-500' : ''}>{todo.title}</span>
      </div>
      <button onClick={handleDelete} className="text-black hover:text-gray-700 transition">
        <Trash2 size={20} />
      </button>
    </li>
  );
}

export default React.memo(TodoItem);
