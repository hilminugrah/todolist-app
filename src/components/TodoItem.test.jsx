import React from 'react';
import { render, screen, fireEvent } from '@testing-library/react';
import { useDispatch } from 'react-redux';
import TodoItem from './TodoItem';
import { toggleTodo, deleteTodo } from '../redux/actions/todoActions';


jest.mock('react-redux', () => ({
  ...jest.requireActual('react-redux'),
  useDispatch: jest.fn(),
}));

jest.mock('../redux/actions/todoActions', () => ({
  toggleTodo: jest.fn((id, completed) => ({ type: 'TOGGLE_TODO', payload: { id, completed } })),
  deleteTodo: jest.fn((id) => ({ type: 'DELETE_TODO', payload: id })),
}));

describe('TodoItem Component', () => {
  let mockDispatch;
  const todo = { id: 1, title: 'Belajar Testing', completed: false };

  beforeEach(() => {
    mockDispatch = jest.fn();
    useDispatch.mockReturnValue(mockDispatch);
    jest.clearAllMocks();
  });

  test('sesuai snapshot', () => {
    const { asFragment } = render(<TodoItem todo={todo} />);
    expect(asFragment()).toMatchSnapshot();
  });

  test('menampilkan judul todo', () => {
    render(<TodoItem todo={todo} />);
    expect(screen.getByText('Belajar Testing')).toBeInTheDocument();
  });

  test('klik checkbox memanggil toggleTodo', () => {
    render(<TodoItem todo={todo} />);

    const checkbox = screen.getByRole('checkbox');
    fireEvent.click(checkbox);

    expect(toggleTodo).toHaveBeenCalledWith(1, true);
    expect(mockDispatch).toHaveBeenCalledWith({
      type: 'TOGGLE_TODO',
      payload: { id: 1, completed: true },
    });
  });

  test('klik tombol hapus memanggil deleteTodo', () => {
    render(<TodoItem todo={todo} />);

    const deleteButton = screen.getByRole('button');
    fireEvent.click(deleteButton);

    expect(deleteTodo).toHaveBeenCalledWith(1);
    expect(mockDispatch).toHaveBeenCalledWith({
      type: 'DELETE_TODO',
      payload: 1,
    });
  });
});
