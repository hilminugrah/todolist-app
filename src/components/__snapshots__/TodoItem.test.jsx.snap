// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`TodoItem Component sesuai snapshot 1`] = `
<DocumentFragment>
  <li
    class="flex justify-between items-center p-4 rounded-lg border border-gray-500 shadow-md bg-white"
  >
    <div
      class="flex items-center space-x-3 relative"
    >
      <input
        class="
            w-5 h-5 appearance-none rounded-full border border-black
            checked:bg-purple-800 checked:border-purple-800
            relative transition duration-200
            after:content-[''] checked:after:content-['âœ”']
            after:absolute after:top-[1px] after:left-[5px]
            checked:after:text-white checked:after:font-bold checked:after:text-sm
          "
        type="checkbox"
      />
      <span
        class=""
      >
        Belajar Testing
      </span>
    </div>
    <button
      class="text-black hover:text-gray-700 transition"
    >
      <svg
        aria-hidden="true"
        class="lucide lucide-trash2 lucide-trash-2"
        fill="none"
        height="20"
        stroke="currentColor"
        stroke-linecap="round"
        stroke-linejoin="round"
        stroke-width="2"
        viewBox="0 0 24 24"
        width="20"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M10 11v6"
        />
        <path
          d="M14 11v6"
        />
        <path
          d="M19 6v14a2 2 0 0 1-2 2H7a2 2 0 0 1-2-2V6"
        />
        <path
          d="M3 6h18"
        />
        <path
          d="M8 6V4a2 2 0 0 1 2-2h4a2 2 0 0 1 2 2v2"
        />
      </svg>
    </button>
  </li>
</DocumentFragment>
`;
